ingressPort.tex

```diff
@@ -515,6 +554,7 @@ \section{Instruction Interface}
     \item \textit{retires}. Maximum number of instructions that can be retired per block is \textit{retires} + 1.
     \item \textit{return\_stack\_size}. Number of entries in the return address stack is 2\textsuperscript{\textbf{return\_stack\_size}}. No stack if 0.
     \item \textit{rv32}. ISA is RV32 when 1.
+    \item \textit{sijump}. Sequentially inferable jumps supported when 1.

```

Comment:
Does the decoder really need to know whether sijumps are supported?  A simpler alternative* is that there can be a requirement that decoders assume that they are supported and it just won't ever see any if they aren't supported.  The encoder is allowed to have a mode that treats all sijumps as uninferable but there's no requirement for the decoder to know that.

\* simpler for hardware, for the hardware-decoder interface, and simpler in some ways for the decoder (less code in a static sense with less discovery and fewer conditionals based on that discovery, though lack of conditionals causes more code to dynamically execute in some cases)

Reply:
No, that won't work.  The CPU, encoder and decoder all have to behave consistently.  

Suppose as you suggest the decoder can support SIJs, and has enabled SIJ mode in the encoder.  However, the CPU does not implement this, and the sijump input is always low.  This means that the encoder will report the target address of all SIJs.

All the decoder receives from the encoder is a format 1/2 packet containing an address, which is usually the target of a jump.  The decoder follows the execution path through the code until it encounters the reported address, or a jump.  It will then deduce that the reported address is the target of this jump.  However, if SIJ mode is enabled, and the decoder sees that the jump is preceded by an AUIPC, it will infer the target of this jump, and keep walking through the code looking for the next jump.  If the encoder had reported the address of the SIJ target (because the CPU hadn't asserted sijump), the encoder and decoder are now out of sync.

The sijump_p parameter on the encoder tells it whether the CPU supports SIJ or not.  If sijump_p is low, and the user tries to enable SIJ mode, the encoder will ignore the request, and the support packet will indicate to the decoder that SIJ mode is disabled.  This will ensure the encoder and decoder remain synced.

I do need to update the pseudo code to add the SIJ mode enable bit though.

Reply:
I see.